Adjust release-process according to experience from kyma 1.4 release (#334)
kyma-project/community: guidelines/releases-guidelines/02-release-process.md

label: internal
---

This document explains how a new Kyma and also new Kyma CLI release can be introduced.

## Kyma

> ### Release Information

Reviewer: Why are we removing this? I think the statement is still valid but the link is broken.

Author: There are dedicated sections for `#### Steps for new major/minor version` and `#### Steps for new release candidate`. Therefore I thought we have it twice if we keep the statement

A Kyma release consists of the following things:
* Docker images for Kyma components
* Github release
  * Release artifacts such as source code and configuration
* Git tag
* Release branch

### Changes to test-infra repository

#### Steps for new major/minor version

Define release jobs on the `master` branch in the `test-infra` repository. To ensure every job name is unique, prefix it with `pre-rel{versionNumber}`. Remember to provide the version number without any periods. For example, to find all jobs for the 1.4 release, look for job names with the `pre-rel14` prefix. To learn how to define a release job for a component, read the following [document](https://github.com/kyma-project/test-infra/blob/master/docs/prow/release-jobs.md).

1. Navigate to the `test-infra` repository.
1. Define release jobs in the `prow/jobs/test-infra` directory in the `watch-pods.yaml` file.

> 1. Define release jobs in the `prow/jobs/kyma` directory in the following files:

Reviewer: Did you intentionally put `1.` for 2 lines? If we intend to execute these steps to be run in parallel, I think dashes should be the way to go. Or `1. 2. 3. ....`

Author: As discussed in person: when using a markdown viewer each `1.` is automatically incremented

	```bash
	echo -n {RELEASE_VERSION} > prow/RELEASE_VERSION
	```

1. Push the branch to the `test-infra` repository.

1. Create a PR to `test-infra/release-x.y`. This triggers the pre-release job for `watch-pods`. 

	> **NOTE:** To trigger the `watch-pods` build without introducing any changes, edit any file within the `test-infra` repository and create a pull request. You don't need to merge it as the job is triggered anyway. After a successful `watch-pods` image build, close the pull request.

1. Update the `RELEASE_VERSION` file with the name of the next minor release candidate and merge the pull request to `master`. For example, if the `RELEASE_VERSION` on the `master` branch is set to `1.4.2`, then change the version to `1.5.0-rc1`.

### Changes to Kyma repository

#### Steps for new major/minor version

1. Create a release branch in the `kyma` repository.

	Do it only for major and minor releases.
	The name of this branch should follow the `release-x.y` pattern, such as `release-1.4`.

	```bash
	git fetch upstream
	git checkout -b release-{release_version} upstream/master
	git push upstream release-{release_version}
	```

> 	> **NOTE:** If you push anything to the release branch (includes creating of the branch), or if you rebase the branch, a new Github release is created. This is because of the post-submit job `post-rel14-kyma-release-candidate` in `kyma-github-release.yaml` and only happens if you already merged the test-infra changes.

Reviewer: Why don't we do this step before creating a release branch in test-infra or merging new release changes to master/test-infra? Because when a developer reaches this step, a false release will anyway get created as there is no point of return. All we need to make sure is not to push the changes in test-infra before creating/rebasing(rarely needed) the release branch of Kyma repo. This can be done on the day of the release rc1. And the preparation work on test-infra can start before(except `git push ....`)

Author: I incorporated your feedback. Please see the new commit

13. Create a PR to `test-infra/release-x.y`. This triggers the pre-release job for `watch-pods`. 

	> **NOTE:** To trigger the `watch-pods` build without introducing any changes, edit any file within the `test-infra` repository and create a pull request. You don't need to merge it as the job is triggered anyway. After a successful `watch-pods` image build, close the pull request.

14. Update the `RELEASE_VERSION` file with the name of the next minor release candidate and merge the pull request to `master`. For example, if the `RELEASE_VERSION` on the `master` branch is set to `1.4.2`, then change the version to `1.5.0-rc1`.

### Changes to Kyma repository

#### Steps for new major/minor version

In  the `resources/core/values.yaml` file, replace the `clusterDocsTopicsVersion` value with your release branch name. For example, for the 1.4.1 release, find the following section:

```yaml
docs:
    # (...) - truncated
    clusterDocsTopicsVersion: master
```

And replace the `clusterDocsTopicsVersion` value with the following:

```yaml
docs:
    # (...)
    clusterDocsTopicsVersion: release-1.4
```
  
#### Steps for new release candidate

> >**NOTE:** This point applies for new release candidates such as rc1, rc2 and the final version.

Reviewer 2: ```suggestion
>**NOTE:** This section applies to new release candidates such as RC1, RC2, and the final release version.
```

		dir:
		version: 1.1.0-rc1
		```

> 		> **CAUTION**: Do **not** update the version of components whose `dir` section does not contain `develop`, as is the case with Console-related components. Also do not change octopus image in `kyma/resources/testing/values.yaml` even though it has `develop`.

Author: Example:
```yaml
global:
  octopus:
    image:
      registry: eu.gcr.io/kyma-project/incubator
      dir: develop/
      version: c919e4ae
```
The user should not change anything here. It is not only not changing the dir. `image` is the parent key, but still not very clear instructions.

```suggestion
		> **CAUTION**: Do **not** update the version of components whose `dir` section does not contain `develop`, as is the case with Console-related components. Also do not change octopus version in `kyma/resources/testing/values.yaml` even though the directory is `develop` .
```

> 		> **CAUTION**: Do **not** update the version of components whose `dir` section does not contain `develop`, as is the case with Console-related components. Also do not change octopus image in `kyma/resources/testing/values.yaml` even though it has `develop`.

Reviewer 2: We can then stick to your version :) It is clearer.

11. Create a spreadsheet with all open issues labeled as `test-missing`. Every team assigned to an issue must cover the outstanding test with manual verification on every release candidate. After the test is finished successfully, the responsible team must mark it as completed in the spreadsheet. Every issue identified during testing must be reported. To make the testing easier, provision a publicly available cluster with the release candidate version after you complete all steps listed in this document.

12. Notify Team Breaking Pixels that the release is available for integration with Faros.

## CLI

A release of `cli` consists of:
* bumping the code to a release tag/branch
* creating artifacts (cli binaries and source code archives)
* github release with automated changelog generation

First the current code is pinned in a release branch.
Then the Kyma version is updated to test the integration with the latest Kyma version. Finally cli is released via a git tag which triggers a Github release.

### Steps for new major/minor version

> >**NOTE:** This point applies only to new major and minor versions and not for new release candidates such as rc2 when already having rc1.

Author: Why two `>>` ?

> >**NOTE:** This point applies only to new major and minor versions and not for new release candidates such as rc2 when already having rc1.

Reviewer 2: My bad :D Should be one of course :)
